src/components/CustomSelect/CustomSelectDropdown/CustomSelectDropdown.tsx

```diff
@@ -0,0 +1,374 @@
+      // apply the nearest parents z-index + 1
+      const zIndex = getNearestParentsZIndex(toggle);
+      if (parseInt(zIndex) > 0) {
+        dropdownRef.current.parentElement?.style.setProperty(
+          "z-index",
+          zIndex + 1,
+        );
+      }
+    }
+
+    setTimeout(() => {
+      if (isSearchable) {
+        searchRef.current?.focus();
+        return;
+      }
+
+      dropdownRef.current?.focus();
+    }, 100);
+  }, [isSearchable, toggleId]);
+
+  const handleResize = () => {
+    adjustDropdownHeight(dropdownListRef.current, searchRef.current);
+  };
+
+  useLayoutEffect(handleResize, []);
+  useListener(window, handleResize, "resize");
+
+  // track selected index from key board action and scroll into view if needed
+  useEffect(() => {
+    optionsRef.current[selectedIndex]?.scrollIntoView({
+      block: "nearest",
+      inline: "nearest",
+    });
+  }, [selectedIndex]);
+
+  const filteredOptions = options?.filter((option) => {
+    if (!search || option.disabled) return true;
+    const searchText = getOptionText(option) || option.value;
+    return searchText.toLowerCase().includes(search);
+  });
+
+  const getNextOptionIndex = (goingUp: boolean, prevIndex: number) => {
+    const increment = goingUp ? -1 : 1;
+    let currIndex = prevIndex + increment;
+    // skip disabled options for key board action
+    while (filteredOptions[currIndex] && filteredOptions[currIndex]?.disabled) {
+      currIndex += increment;
+    }
+
+    // consider upper bound for navigating down the list
+    if (increment > 0) {
+      return currIndex < filteredOptions.length ? currIndex : prevIndex;
+    }
+
+    // consider lower bound for navigating up the list
+    return currIndex >= 0 ? currIndex : prevIndex;
+  };
+
+  // handle keyboard actions for navigating the select dropdown
+  const handleKeyDown = (event: KeyboardEvent<HTMLDivElement>) => {
+    const upDownKeys = ["ArrowUp", "ArrowDown"];
+
+    // prevent default browser actions for up, down, enter and escape keys
+    // also prevent any other event listeners from being called up the DOM tree
+    if ([...upDownKeys, "Enter", "Escape", "Tab"].includes(event.key)) {
+      event.preventDefault();
+      event.nativeEvent.stopImmediatePropagation();
+    }
+
+    if (upDownKeys.includes(event.key)) {
+      setSelectedIndex((prevIndex) => {
+        const goingUp = event.key === "ArrowUp";
+        return getNextOptionIndex(goingUp, prevIndex);
+      });
+    }
+
+    if (event.key === "Enter" && filteredOptions[selectedIndex]) {
+      onSelect(filteredOptions[selectedIndex].value);
+    }
+
+    if (event.key === "Escape" || event.key === "Tab") {
+      onClose();
+    }
+  };
+
+  const handleSearch = (value: string) => {
+    setSearch(value.toLowerCase());
+    // reset selected index when search text changes
+    setSelectedIndex(0);
+    optionsRef.current = [];
+  };
```

Comment:
How much work will it be to take an exposed `onSearch` prop and call it here with the value every time the search changes?

```suggestion
  const handleSearch = (value: string) => {
    setSearch(value.toLowerCase());
    
    if (onSearch) {
      onSearch(value);
      return;
    }
    
    // reset selected index when search text changes
    setSelectedIndex(0);
    optionsRef.current = [];
  };
```

This could be useful if we want to update the list of options asynchronously instead of simply filtering the existing options based on the search value. We could use it alongside a `searchExternallyControlled` or similar prop.

Reply:
Good suggestion, it's not that much effort and even so, if it's an good idea for improvement we should implement it :smile: will address this

Reply:
Awesome! Don't mind the return statement in the suggestion. I thought the "setSelectedIndex" was for updating the selected option but this is only for updating the keyboard navigation.

Reply:
sure, I think the ordering of the code will need to shuffle a bit in the suggestion. `setSelectedIndex` and the reset for `optionsRef` should execute before the return, and `setSearch` should not execute if `onSearch` is defined since we do not want internal filtering to happen at all, the filtering will be delegated to the parent and passed down via the `options` prop.

Additionally, we can also expose `setSelectedIndex` via `onSearch` for the parent component to decide which option to highlight upon search. Do you think this is a plausible use case?

Reply:
We don't need to expose `setSelectedIndex`. If the options are externally controlled through search, they would most likely change every time the user searches something. if we just keep it as `setSelectedIndex(0)`, it will naturally highlight the top option until there is only one option.

Reply:
Yes, `setActiveOptionIndex` sounds good, or maybe we can do `setHighlightedOptionIndex`, since active can still mean selected. wdyt?

Reply:
Yeah initially thought about `setHighlightedOptionIndex` but felt it was a bit verbose. Happy to use that instead since it seems like a more popular choice :+1:

Reply:
I've made the adjustment, please have a look.

Reply:
Looks great, thank you!